org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/hcr/ChangedClassFilesVisitor.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2015 Contributor. All rights reserved.
+ */
+package org.scalaide.debug.internal.hcr
+
+import org.eclipse.core.resources.IFile
+import org.eclipse.core.resources.IMarker
+import org.eclipse.core.resources.IResource
+import org.eclipse.core.resources.IResourceDelta
+import org.eclipse.core.resources.IResourceDeltaVisitor
+import org.eclipse.core.runtime.IPath
+import org.eclipse.core.runtime.Path
+import org.eclipse.jdt.core.ICompilationUnit
+import org.eclipse.jdt.core.IJavaModelMarker
+import org.eclipse.jdt.core.IJavaProject
+import org.eclipse.jdt.core.JavaCore
+import org.eclipse.jdt.core.ToolFactory
+import org.eclipse.jdt.core.util.IClassFileReader
+import org.eclipse.jdt.internal.debug.core.JavaDebugUtils
+import org.scalaide.logging.HasLogger
+import org.scalaide.debug.internal.preferences.HotCodeReplacePreferences
+
+private[internal] case class ClassFileResource(fullyQualifiedName: String, classFile: IFile)
+
+/**
+ * Collects classes affected by changes in code and optionally skips these ones which contain
+ * error markers in associated source files.
+ * The implementation of this class is inspired by
+ * org.eclipse.jdt.internal.debug.core.hcr.JavaHotCodeReplaceManager.ChangedClassFilesVisitor
+ * used in Java HCR implementation.
+ */
+private[internal] class ChangedClassFilesVisitor extends IResourceDeltaVisitor with HasLogger {
+  import ChangedClassFilesVisitor._
+
+  // the value stored as a field to finish applying the same strategy to all classes even when someone changed
+  // the configuration in the meantime
+  private var replaceDespiteCompilationErrors = HotCodeReplacePreferences.performHcrForFilesContainingErrors
+
+  /**
+   * Found classes containing changes.
+   */
+  private val changedClasses = scala.collection.mutable.Set[ClassFileResource]()
+
+  def getChangedClasses: List[ClassFileResource] = changedClasses.toList
+
+  /**
+   * It should be called always before the visitor is reused.
+   */
+  def reset(): Unit = {
```

Comment:
I don't really like the mutability in this class. I find it simpler to make the class "one-time-use" than to remember to call reset from time to time. Clients simply instantiate another one (there doesn't seem to be any expensive operation in the constructor, so the cost is negligible).